kvstore/store.py

```diff
@@ -1,27 +1,77 @@
 import os
 import pickle
 from kvstore.constants import NULL
+from kvstore.encoding import WriteLog, BinaryEncoderDecoder, Set, Snapshot
 
-DEFAULT_FILE_NAME = "data/store.p"
+STORE_FILE_TMPL = "data/store%s.p"
+WRITE_LOG_TMPL = "data/writelog%s.p"
 
 class KVStore:
-    def __init__(self, filename=DEFAULT_FILE_NAME):
-        self.filename = filename
-        self.store = self.read_from_disk()
+    def __init__(self, node_number, filename=STORE_FILE_TMPL):
+        self.en = BinaryEncoderDecoder()
 
-    def write_to_disk(self):
-        pickle.dump(self.store, open( self.filename, "wb+"))
+        self.node_number = node_number
+        self.filename = filename % node_number
+        self.writelog = WRITE_LOG_TMPL % node_number
+
+        read = self.read_from_disk()
+        self.store = read["store"]
+        self.logSequenceNumber = read["logSequenceNumber"]
+
+    def write_to_disk(self, command):
+        # update log sequence number
+        self.logSequenceNumber += 1
+        print(f"logSequenceNumber: {self.logSequenceNumber}")
+
+        self.dump_db()
+
+        # write to write log
+        wl = WriteLog(command.key, command.value, self.logSequenceNumber)
+        encoded_wl = self.en.encode(wl)
+        with open(self.writelog, 'ab') as f:
+            f.write(encoded_wl)
+        return wl
+
+    def dump_db(self):
+        pickle.dump(
+            {"store": self.store, "logSequenceNumber": self.logSequenceNumber},
+            open( self.filename, "wb+")
+        )
 
     def read_from_disk(self):
         if os.path.exists(self.filename):
-            return pickle.load( open( self.filename, "rb+" ) )
+            fetched = pickle.load( open( self.filename, "rb+" ) )
+            if fetched.get("logSequenceNumber"):
+                return fetched
+            return { "store": fetched, "logSequenceNumber": 0 }
         else:
-            return {}
+            return {
+                "store": {},
+                "logSequenceNumber": 0
+            }
 
     def get(self, key):
         return self.store.get(key, NULL)
 
+    def start_from_snapshot(self, snapshot):
+        print("loading from snapshot: ", snapshot)
+        self.logSequenceNumber = snapshot.logSequenceNumber
+        self.store = snapshot.store
+        self.dump_db()
+
+    def get_snapshot(self):
+        return self.store, self.logSequenceNumber
+
     def set(self, key, value):
         self.store[key] = value
-        self.write_to_disk()
-        return value
+        wl = self.write_to_disk(Set(key, value))
+        return wl
+
+    def receive_write_log(self, command):
+        if command.logSequenceNumber < self.logSequenceNumber:
+            print("received old write. ignoring...")
+        elif command.logSequenceNumber > self.logSequenceNumber + 1:
+            # TODO: handle writes that have bene received out of order
```

Comment:
to work around needing to send writes in order, my server sends the sequence number to start at and a list of all writes that have been made since that number.